### PR TITLE
MULE-15325: BindingContext look up is showing significant performance

### PR DESCRIPTION
**Pull request by elrodro83**
overhead

Baseline:
```
BindingContextBenchmark.fiveBindings                                     avgt   10    60.302 ±   1.825   ns/op
BindingContextBenchmark.fiveBindings:·gc.alloc.rate                      avgt   10  2169.141 ±  66.248  MB/sec
BindingContextBenchmark.fiveBindings:·gc.alloc.rate.norm                 avgt   10   144.000 ±   0.001    B/op
BindingContextBenchmark.fiveBindings:·gc.churn.PS_Eden_Space             avgt   10  2169.684 ±  71.569  MB/sec
BindingContextBenchmark.fiveBindings:·gc.churn.PS_Eden_Space.norm        avgt   10   144.034 ±   1.122    B/op
BindingContextBenchmark.fiveBindings:·gc.churn.PS_Survivor_Space         avgt   10     0.158 ±   0.041  MB/sec
BindingContextBenchmark.fiveBindings:·gc.churn.PS_Survivor_Space.norm    avgt   10     0.010 ±   0.003    B/op
BindingContextBenchmark.fiveBindings:·gc.count                           avgt   10   997.000            counts
BindingContextBenchmark.fiveBindings:·gc.time                            avgt   10   867.000                ms
BindingContextBenchmark.fromEvent                                      avgt   10   467.912 ±  52.445   ns/op
BindingContextBenchmark.fromEvent:·gc.alloc.rate                       avgt   10  2448.303 ± 248.480  MB/sec
BindingContextBenchmark.fromEvent:·gc.alloc.rate.norm                  avgt   10  1256.001 ±   0.001    B/op
BindingContextBenchmark.fromEvent:·gc.churn.PS_Eden_Space              avgt   10  2447.995 ± 247.201  MB/sec
BindingContextBenchmark.fromEvent:·gc.churn.PS_Eden_Space.norm         avgt   10  1255.863 ±   7.504    B/op
BindingContextBenchmark.fromEvent:·gc.churn.PS_Survivor_Space          avgt   10     0.177 ±   0.049  MB/sec
BindingContextBenchmark.fromEvent:·gc.churn.PS_Survivor_Space.norm     avgt   10     0.091 ±   0.020    B/op
BindingContextBenchmark.fromEvent:·gc.count                            avgt   10  1116.000            counts
BindingContextBenchmark.fromEvent:·gc.time                             avgt   10   904.000                ms
BindingContextBenchmark.parentLookup                                   avgt   10    23.830 ±   0.440   ns/op
BindingContextBenchmark.parentLookup:·gc.alloc.rate                    avgt   10   609.816 ±  10.951  MB/sec
BindingContextBenchmark.parentLookup:·gc.alloc.rate.norm               avgt   10    16.000 ±   0.001    B/op
BindingContextBenchmark.parentLookup:·gc.churn.PS_Eden_Space           avgt   10   610.434 ±  12.042  MB/sec
BindingContextBenchmark.parentLookup:·gc.churn.PS_Eden_Space.norm      avgt   10    16.016 ±   0.083    B/op
BindingContextBenchmark.parentLookup:·gc.churn.PS_Survivor_Space       avgt   10     0.136 ±   0.027  MB/sec
BindingContextBenchmark.parentLookup:·gc.churn.PS_Survivor_Space.norm  avgt   10     0.004 ±   0.001    B/op
BindingContextBenchmark.parentLookup:·gc.count                         avgt   10  1150.000            counts
BindingContextBenchmark.parentLookup:·gc.time                          avgt   10   883.000                ms
BindingContextBenchmark.payloadBinding                                   avgt   10     9.458 ±   0.353   ns/op
BindingContextBenchmark.payloadBinding:·gc.alloc.rate                    avgt   10  1536.240 ±  56.332  MB/sec
BindingContextBenchmark.payloadBinding:·gc.alloc.rate.norm               avgt   10    16.000 ±   0.001    B/op
BindingContextBenchmark.payloadBinding:·gc.churn.PS_Eden_Space           avgt   10  1536.608 ±  54.532  MB/sec
BindingContextBenchmark.payloadBinding:·gc.churn.PS_Eden_Space.norm      avgt   10    16.004 ±   0.109    B/op
BindingContextBenchmark.payloadBinding:·gc.churn.PS_Survivor_Space       avgt   10     0.131 ±   0.018  MB/sec
BindingContextBenchmark.payloadBinding:·gc.churn.PS_Survivor_Space.norm  avgt   10     0.001 ±   0.001    B/op
BindingContextBenchmark.payloadBinding:·gc.count                         avgt   10  1147.000            counts
BindingContextBenchmark.payloadBinding:·gc.time                          avgt   10   909.000                ms
BindingContextBenchmark.threeBindings                                    avgt   10    37.404 ±   1.763   ns/op
BindingContextBenchmark.threeBindings:·gc.alloc.rate                     avgt   10  2526.336 ± 121.390  MB/sec
BindingContextBenchmark.threeBindings:·gc.alloc.rate.norm                avgt   10   104.000 ±   0.001    B/op
BindingContextBenchmark.threeBindings:·gc.churn.PS_Eden_Space            avgt   10  2523.548 ± 118.576  MB/sec
BindingContextBenchmark.threeBindings:·gc.churn.PS_Eden_Space.norm       avgt   10   103.889 ±   0.843    B/op
BindingContextBenchmark.threeBindings:·gc.churn.PS_Survivor_Space        avgt   10     0.181 ±   0.026  MB/sec
BindingContextBenchmark.threeBindings:·gc.churn.PS_Survivor_Space.norm   avgt   10     0.007 ±   0.001    B/op
BindingContextBenchmark.threeBindings:·gc.count                          avgt   10  1167.000            counts
BindingContextBenchmark.threeBindings:·gc.time                           avgt   10   907.000                ms
BindingContextBenchmark.varsBinding                                      avgt   10     9.719 ±   0.323   ns/op
BindingContextBenchmark.varsBinding:·gc.alloc.rate                       avgt   10  1495.065 ±  48.062  MB/sec
BindingContextBenchmark.varsBinding:·gc.alloc.rate.norm                  avgt   10    16.000 ±   0.001    B/op
BindingContextBenchmark.varsBinding:·gc.churn.PS_Eden_Space              avgt   10  1495.857 ±  52.260  MB/sec
BindingContextBenchmark.varsBinding:·gc.churn.PS_Eden_Space.norm         avgt   10    16.008 ±   0.074    B/op
BindingContextBenchmark.varsBinding:·gc.churn.PS_Survivor_Space          avgt   10     0.128 ±   0.031  MB/sec
BindingContextBenchmark.varsBinding:·gc.churn.PS_Survivor_Space.norm     avgt   10     0.001 ±   0.001    B/op
BindingContextBenchmark.varsBinding:·gc.count                            avgt   10  1116.000            counts
BindingContextBenchmark.varsBinding:·gc.time                             avgt   10   885.000                ms
BindingContextBenchmark.withParent                                     avgt   10   275.382 ±   9.090   ns/op
BindingContextBenchmark.withParent:·gc.alloc.rate                      avgt   10  2981.929 ±  94.124  MB/sec
BindingContextBenchmark.withParent:·gc.alloc.rate.norm                 avgt   10   904.001 ±   0.001    B/op
BindingContextBenchmark.withParent:·gc.churn.PS_Eden_Space             avgt   10  2981.156 ± 100.782  MB/sec
BindingContextBenchmark.withParent:·gc.churn.PS_Eden_Space.norm        avgt   10   903.743 ±   3.778    B/op
BindingContextBenchmark.withParent:·gc.churn.PS_Survivor_Space         avgt   10     0.157 ±   0.031  MB/sec
BindingContextBenchmark.withParent:·gc.churn.PS_Survivor_Space.norm    avgt   10     0.048 ±   0.009    B/op
BindingContextBenchmark.withParent:·gc.count                           avgt   10  1196.000            counts
BindingContextBenchmark.withParent:·gc.time                            avgt   10   927.000                ms
```

After (including https://github.com/mulesoft/mule-api/pull/493):
```
BindingContextBenchmark.fiveBindings                                    avgt   10    55.351 ±   3.687   ns/op
BindingContextBenchmark.fiveBindings:·gc.alloc.rate                     avgt   10  1840.294 ± 114.409  MB/sec
BindingContextBenchmark.fiveBindings:·gc.alloc.rate.norm                avgt   10   112.000 ±   0.001    B/op
BindingContextBenchmark.fiveBindings:·gc.churn.PS_Eden_Space            avgt   10  1841.024 ± 108.107  MB/sec
BindingContextBenchmark.fiveBindings:·gc.churn.PS_Eden_Space.norm       avgt   10   112.056 ±   0.916    B/op
BindingContextBenchmark.fiveBindings:·gc.churn.PS_Survivor_Space        avgt   10     0.172 ±   0.035  MB/sec
BindingContextBenchmark.fiveBindings:·gc.churn.PS_Survivor_Space.norm   avgt   10     0.010 ±   0.002    B/op
BindingContextBenchmark.fiveBindings:·gc.count                          avgt   10  1155.000            counts
BindingContextBenchmark.fiveBindings:·gc.time                           avgt   10   918.000                ms
BindingContextBenchmark.fromEvent                                       avgt   10   201.094 ±  26.584   ns/op
BindingContextBenchmark.fromEvent:·gc.alloc.rate                        avgt   10  3051.030 ± 336.195  MB/sec
BindingContextBenchmark.fromEvent:·gc.alloc.rate.norm                   avgt   10   672.001 ±   0.001    B/op
BindingContextBenchmark.fromEvent:·gc.churn.PS_Eden_Space               avgt   10  3049.492 ± 347.785  MB/sec
BindingContextBenchmark.fromEvent:·gc.churn.PS_Eden_Space.norm          avgt   10   671.529 ±   4.619    B/op
BindingContextBenchmark.fromEvent:·gc.churn.PS_Survivor_Space           avgt   10     0.160 ±   0.037  MB/sec
BindingContextBenchmark.fromEvent:·gc.churn.PS_Survivor_Space.norm      avgt   10     0.035 ±   0.007    B/op
BindingContextBenchmark.fromEvent:·gc.count                             avgt   10  1069.000            counts
BindingContextBenchmark.fromEvent:·gc.time                              avgt   10   879.000                ms
BindingContextBenchmark.parentLookup                                    avgt   10    22.085 ±   0.583   ns/op
BindingContextBenchmark.parentLookup:·gc.alloc.rate                     avgt   10   658.001 ±  16.954  MB/sec
BindingContextBenchmark.parentLookup:·gc.alloc.rate.norm                avgt   10    16.000 ±   0.001    B/op
BindingContextBenchmark.parentLookup:·gc.churn.PS_Eden_Space            avgt   10   658.655 ±  15.255  MB/sec
BindingContextBenchmark.parentLookup:·gc.churn.PS_Eden_Space.norm       avgt   10    16.017 ±   0.111    B/op
BindingContextBenchmark.parentLookup:·gc.churn.PS_Survivor_Space        avgt   10     0.134 ±   0.022  MB/sec
BindingContextBenchmark.parentLookup:·gc.churn.PS_Survivor_Space.norm   avgt   10     0.003 ±   0.001    B/op
BindingContextBenchmark.parentLookup:·gc.count                          avgt   10  1130.000            counts
BindingContextBenchmark.parentLookup:·gc.time                           avgt   10   882.000                ms
BindingContextBenchmark.payloadBinding                                  avgt   10     4.482 ±   0.214   ns/op
BindingContextBenchmark.payloadBinding:·gc.alloc.rate                   avgt   10     0.003 ±   0.001  MB/sec
BindingContextBenchmark.payloadBinding:·gc.alloc.rate.norm              avgt   10    ≈ 10⁻⁵              B/op
BindingContextBenchmark.payloadBinding:·gc.count                        avgt   10       ≈ 0            counts
BindingContextBenchmark.threeBindings                                   avgt   10    18.613 ±   0.315   ns/op
BindingContextBenchmark.threeBindings:·gc.alloc.rate                    avgt   10  3512.378 ±  57.318  MB/sec
BindingContextBenchmark.threeBindings:·gc.alloc.rate.norm               avgt   10    72.000 ±   0.001    B/op
BindingContextBenchmark.threeBindings:·gc.churn.PS_Eden_Space           avgt   10  3511.613 ±  64.366  MB/sec
BindingContextBenchmark.threeBindings:·gc.churn.PS_Eden_Space.norm      avgt   10    71.984 ±   0.392    B/op
BindingContextBenchmark.threeBindings:·gc.churn.PS_Survivor_Space       avgt   10     0.188 ±   0.022  MB/sec
BindingContextBenchmark.threeBindings:·gc.churn.PS_Survivor_Space.norm  avgt   10     0.004 ±   0.001    B/op
BindingContextBenchmark.threeBindings:·gc.count                         avgt   10  1216.000            counts
BindingContextBenchmark.threeBindings:·gc.time                          avgt   10   933.000                ms
BindingContextBenchmark.varsBinding                                     avgt   10     6.685 ±   0.023   ns/op
BindingContextBenchmark.varsBinding:·gc.alloc.rate                      avgt   10  2172.609 ±   7.465  MB/sec
BindingContextBenchmark.varsBinding:·gc.alloc.rate.norm                 avgt   10    16.000 ±   0.001    B/op
BindingContextBenchmark.varsBinding:·gc.churn.PS_Eden_Space             avgt   10  2172.114 ±  20.516  MB/sec
BindingContextBenchmark.varsBinding:·gc.churn.PS_Eden_Space.norm        avgt   10    15.996 ±   0.136    B/op
BindingContextBenchmark.varsBinding:·gc.churn.PS_Survivor_Space         avgt   10     0.149 ±   0.024  MB/sec
BindingContextBenchmark.varsBinding:·gc.churn.PS_Survivor_Space.norm    avgt   10     0.001 ±   0.001    B/op
BindingContextBenchmark.varsBinding:·gc.count                           avgt   10  1186.000            counts
BindingContextBenchmark.varsBinding:·gc.time                            avgt   10   916.000                ms
BindingContextBenchmark.withParent                                      avgt   10    78.876 ±   1.494   ns/op
BindingContextBenchmark.withParent:·gc.alloc.rate                       avgt   10  4421.368 ±  81.864  MB/sec
BindingContextBenchmark.withParent:·gc.alloc.rate.norm                  avgt   10   384.000 ±   0.001    B/op
BindingContextBenchmark.withParent:·gc.churn.PS_Eden_Space              avgt   10  4422.621 ±  67.967  MB/sec
BindingContextBenchmark.withParent:·gc.churn.PS_Eden_Space.norm         avgt   10   384.120 ±   2.260    B/op
BindingContextBenchmark.withParent:·gc.churn.PS_Survivor_Space          avgt   10     0.175 ±   0.028  MB/sec
BindingContextBenchmark.withParent:·gc.churn.PS_Survivor_Space.norm     avgt   10     0.015 ±   0.002    B/op
BindingContextBenchmark.withParent:·gc.count                            avgt   10  1205.000            counts
BindingContextBenchmark.withParent:·gc.time                             avgt   10   935.000                ms
```

### CODE REVIEW

File: core/src/main/java/org/mule/runtime/core/internal/el/DefaultBindingContextBuilder.java [hunk 1]
```diff
@@ -21,87 +25,194 @@
...
+  private Supplier<TypedValue> varsBinding;
+
+  private final Map<String, Supplier<TypedValue>> bindings = new HashMap<>();
+  private Collection<ExpressionModule> modules = null;
 
   public DefaultBindingContextBuilder() {
-    this.bindings = new HashMap<>();
-    this.modules = new ArrayList<>();
+    this.parent = NULL_BINDING_CONTEXT;
+    // this.bindings = new HashMap<>();
```
Comment (afelisatti) [inline core/src/main/java/org/mule/runtime/core/internal/el/DefaultBindingContextBuilder.java]: delete

File: core/src/main/java/org/mule/runtime/core/internal/el/DefaultBindingContextBuilder.java [hunk 2]
```diff
@@ -21,87 +25,194 @@
...
+    }
     return this;
   }
 
   @Override
   public BindingContext.Builder addBinding(String identifier, Supplier<TypedValue> lazyValue) {
-    bindings.put(identifier, lazyValue);
+    switch (identifier) {
+      case PAYLOAD:
+        payloadBinding = lazyValue.get();
```
Comment (afelisatti) [inline core/src/main/java/org/mule/runtime/core/internal/el/DefaultBindingContextBuilder.java]: Why can't this have lazy values too?

File: core/src/main/java/org/mule/runtime/core/internal/el/DefaultBindingContextBuilder.java [hunk 3]
```diff
@@ -21,87 +25,194 @@
...
+        break;
+    }
     return this;
   }
 
   @Override
   public BindingContext.Builder addAll(BindingContext context) {
-    context.identifiers().forEach(id -> bindings.put(id, () -> context.lookup(id).get()));
-    modules.addAll(context.modules());
+    if (parent == null) {
```
Comment (afelisatti) [inline core/src/main/java/org/mule/runtime/core/internal/el/DefaultBindingContextBuilder.java]: If we were overriding the values before then this does not seem accurate.

File: core/src/main/java/org/mule/runtime/core/internal/el/DefaultBindingContextBuilder.java [hunk 4]
```diff
@@ -21,87 +25,194 @@
...
+      parent = context;
+
+      payloadBinding = context.lookup(PAYLOAD).orElse(null);
+      attributesBinding = context.lookup(ATTRIBUTES).orElse(null);
+      varsBinding = () -> context.lookup(VARS).orElse(null);
+    } else {
+      // context.identifiers().forEach(id -> bindings.put(id, () -> context.lookup(id).get()));
```
Comment (afelisatti) [inline core/src/main/java/org/mule/runtime/core/internal/el/DefaultBindingContextBuilder.java]: ?

File: core/src/main/java/org/mule/runtime/core/internal/el/DefaultBindingContextBuilder.java (hunk 2 above)
Comment (elrodro83) [reply core/src/main/java/org/mule/runtime/core/internal/el/DefaultBindingContextBuilder.java]: the lazy values optimization is for those bindings for which a `TypedValue` is built. `payload` and `attributes` are already `TypedValues`, so no need to make them lazy.

**Pull request merged**